File: src/main/java/com/carry_laundry/carry_backend/config/S3Config.java
package com.carry_laundry.carry_backend.config;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3AsyncClient;
import software.amazon.awssdk.services.s3.S3Configuration;

@Configuration
public class S3Config {

    @Bean
    public S3AsyncClient s3AsyncClient(@Value("${aws.s3.access-key}") String accessKey,
        @Value("${aws.s3.secret-key}") String secretKey) {
        AwsBasicCredentials awsBasicCredentials = AwsBasicCredentials.create(accessKey, secretKey);
        return S3AsyncClient.builder()
            .region(Region.AP_NORTHEAST_2)
            .credentialsProvider(StaticCredentialsProvider.create(awsBasicCredentials))
            .serviceConfiguration(S3Configuration.builder()
                .checksumValidationEnabled(false)
                .chunkedEncodingEnabled(true)
                .build())
            .build();
    }
}

File: carry-infra-s3/src/main/kotlin/com/carry/infra/s3/S3Config.kt
package com.carry.infra.s3

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.s3.S3Client
import java.net.URI

@Configuration
class S3Config {

    @Value("\${cloud.aws.s3.endpoint:}")
    private lateinit var endpoint: String

    @Value("\${cloud.aws.credentials.access-key:}")
    private lateinit var accessKey: String

    @Value("\${cloud.aws.credentials.secret-key:}")
    private lateinit var secretKey: String

    @Value("\${cloud.aws.region.static:ap-northeast-2}")
    private lateinit var region: String

    @Bean
    fun s3Client(): S3Client {
        val builder = S3Client.builder()
            .region(Region.of(region))

        if (accessKey.isNotBlank() && secretKey.isNotBlank()) {
            builder.credentialsProvider(
                StaticCredentialsProvider.create(
                    AwsBasicCredentials.create(accessKey, secretKey)
                )
            )
        }
        if (endpoint.isNotBlank()) {
            builder.endpointOverride(URI.create(endpoint))
                .forcePathStyle(true)
        }
        return builder.build()
    }
}
